comms/src/control_service/worker.rs

```diff
@@ -187,82 +189,278 @@ where
-        CommsCipher::open_with_integral_nonce(&body, &ecdh_shared_secret).map_err(ControlServiceError::CipherError)
+    fn handle_ping(&self, envelope_header: MessageEnvelopeHeader, identity_frame: Frame) -> Result<()> {
+        debug!(target: LOG_TARGET, "Got ping message");
+        self.send_reply(
+            &envelope_header.source,
+            identity_frame,
+            ControlServiceMessageType::Pong,
+            Pong {},
+        )
     }
 
-    fn process_socket_event(&mut self, event: SocketEvent) -> Result<()> {
-        use SocketEventType::*;
+    fn handle_request_connection(
+        &self,
+        envelope_header: MessageEnvelopeHeader,
+        identity_frame: Frame,
+        message: RequestConnection,
+    ) -> Result<()>
+    {
+        debug!(
+            target: LOG_TARGET,
+            "RequestConnection message received (node_id={}, address={})", message.node_id, message.address
+        );
+
+        let pm = &self.connection_manager.peer_manager();
+        let public_key = &envelope_header.source;
+        let peer = match pm.find_with_public_key(&public_key) {
+            Ok(peer) => {
+                if peer.is_banned() {
+                    return Err(ControlServiceError::PeerBanned);
+                }
 
-        match event.event_type {
-            Listening => {
-                info!(target: LOG_TARGET, "Started listening on '{}'", event.address);
-            },
-            Accepted => {
-                info!(target: LOG_TARGET, "Accepted connection on '{}'", event.address);
+                pm.add_net_address(&peer.node_id, &message.control_service_address)
+                    .map_err(ControlServiceError::PeerManagerError)?;
+
+                peer
             },
-            Disconnected => {
-                info!(target: LOG_TARGET, "'{}' Disconnected", event.address);
+            Err(PeerManagerError::PeerNotFoundError) => {
+                let node_id = &message.node_id;
+
+                let peer = Peer::new(
+                    public_key.clone(),
+                    node_id.clone(),
+                    message.control_service_address.clone().into(),
+                    PeerFlags::empty(),
+                );
+
+                pm.add_peer(peer.clone())
+                    .map_err(ControlServiceError::PeerManagerError)?;
+                peer
             },
-            BindFailed => {
-                self.is_running = false;
-                error!(
+            Err(err) => return Err(ControlServiceError::PeerManagerError(err)),
+        };
+
+        // TODO: SECURITY The node ID is not a verified value at this point (PeerNotFoundError branch above).
+        //       An attacker can insert any node id they want to get information about other peers connections
+        //       to this node. For instance, if they already have an active connection.
+        //       The public key should be used as that is validated by the message signature.
```

Comment:
I agree, we need a mechanism to verify or validate node_ids.

Reply:
Think that validating that the node id is the hash of the public key for base nodes will work - then add a different validation for DAN nodes later - so this issue should be easily resolved